homeassistant/components/sensor/habitica.py

```diff
@@ -0,0 +1,151 @@
+"""
+The Habitica sensor.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.habitica/
+"""
+
+import logging
+from collections import namedtuple
+# from typing import Any
+from datetime import timedelta
+import voluptuous as vol
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers import config_validation as cv
+from homeassistant.util import Throttle
+
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = "habitica"
+SENSOR_DATA_IDX = "sensor." + DOMAIN
+DEPENDENCIES = [DOMAIN]
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=15)
+
+CONF_NAME = "name"
+CONF_SENSORS_LIST = "sensors"
+ST = SensorType = namedtuple('SensorType', [
+    "name", "icon", "unit", "path"
+])
+
+SENSORS_TYPES = {
+    'name': ST('Name', None, '', ["profile", "name"]),
+    'hp': ST('HP', 'mdi:heart', 'HP', ["stats", "hp"]),
+    'maxHealth': ST('max HP', 'mdi:heart', 'HP', ["stats", "maxHealth"]),
+    'mp': ST('Mana', 'mdi:auto-fix', 'MP', ["stats", "mp"]),
+    'maxMP': ST('max Mana', 'mdi:auto-fix', 'MP', ["stats", "maxMP"]),
+    'exp': ST('EXP', 'mdi:star', 'EXP', ["stats", "exp"]),
+    'toNextLevel': ST(
+        'Next Lvl', 'mdi:star', 'EXP', ["stats", "toNextLevel"]),
+    'lvl': ST(
+        'Lvl', 'mdi:arrow-up-bold-circle-outline', 'Lvl', ["stats", "lvl"]),
+    'gp': ST('Gold', 'mdi:coin', 'Gold', ["stats", "gp"]),
+    'class': ST('Class', 'mdi:sword', '', ["stats", "class"])
+}
+ALL_SENSORS_TYPES = list(SENSORS_TYPES.keys())
+
+
+def ensure_list_of_option_types(value):
+    """Wrap value in list if it is not one."""
+    if value is None:
+        return ALL_SENSORS_TYPES
+    value = value if isinstance(value, list) else [value]
+    for item in value:
+        if item not in ALL_SENSORS_TYPES:
+            raise vol.Invalid(
+                "Found {}, expected one of {}".format(
+                    item, ALL_SENSORS_TYPES))
+    return value
+
+
+PLATFORM_SCHEMA = cv.PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_NAME): cv.string,
+    vol.Optional(CONF_SENSORS_LIST): ensure_list_of_option_types
+})
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the habitica platform."""
+    name = config.get(CONF_NAME, None)
+    sensors = config.get(CONF_SENSORS_LIST, ALL_SENSORS_TYPES)
+    if name is None:
+        _LOGGER.error("Check name config in Home Assistant config")
+        return False
+    if name not in hass.data[DOMAIN]:
```

Comment:
It's a bit weird that the user needs to coordinate two config sections and make sure that the correct name is in both. Why can't we just setup sensors for all names that the component have setup?

Reply:
Not in both. `name` is optional in component config. It is deduced from api.
The user may not want to configure sensors for all habitica users.

Reply:
Yes but still. We want to be able to validate the user config. This approach makes that hard.

Reply:
Ok. So I'm going to rewrite my patch like this:
* By default all sensors from `ALL_SENSOR_TYPES` are initialized for all `name`s defined in platform.
* User can exclude `name`s from generation via `exclude_names` list
* User can whitelist and blacklist sensors from being generated via `sensors_customize`
So a new example configuration would look like this:
```yaml
habitica:
   - api_user: JOHNDOE-ID # name: johndoe
     api_key: JOHNDOE-PASS
  
   - api_user: MARIADOE-ID # name: mariadoe
     api_key: MARIADOE-PASS
 
   - api_user: EXNICK-ID # name: exnick
     api_key: EXNICK-PASS

sensor:
    - platform: habitica
      exclude_names: ["exnick"]
      sensors_customize:
          johndoe:
              include_only: ["name", "hp", "mp", "toNextLevel", "gp"]
          mariadoe:
              exclude: ["maxMP"]
``` 
`include_only` and `exclude` key names are subject to discussion.
We can generate a log warning if a certain `name` from `exclude_names` or `sensors_customize` keys is missing from component config.

Reply:
Are there any downsides to creating a sensor entity for a username that the user isn't interested in? The user can still decide no to use the entity and to hide it in the frontend etc.

The new proposal is still hard to validate since we don't know the usernames in advance. I suggest to create entities for all usernames that have been configured by the component. The sensor types can have a whitelist (monitored_conditions) since we know the possible sensor types in advance.